File: python/structural/flyweight/map/icon_factory.py
from map.icon import Icon

class IconFactory:
    _instance = None
    _icons: dict = {}

    def __new__(cls):
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def get_icon(self, icon_type: str) -> Icon:
        if icon_type not in self._icons:
            print(f"🆕 Creating new icon for: {icon_type}")
            path = f"/icons/{icon_type}.png"
            self._icons[icon_type] = Icon(icon_type, path)
        else:
            print(f"✅ Reusing icon for: {icon_type}")

        return self._icons[icon_type]
    

File: typescript/src/structural/flyweight/map/IconFactory.ts
import { Icon } from "./Icon";

export class IconFactory{
    private static instance: IconFactory;
    private icons: Map<string, Icon>

    private constructor(){
        this.icons = new Map<string, Icon>()
    }

    public static getInstance(): IconFactory {
        if(!IconFactory.instance){
            IconFactory.instance = new IconFactory();
        }
        return IconFactory.instance;
    }

    getIcon(iconType: string): Icon {
        if (!this.icons.has(iconType)){
            console.log(`🆕 Creating new icon for: ${iconType}`)
            let path: string = `/icons/${iconType}.png`
            const icon: Icon = new Icon(iconType, path)
            this.icons.set(iconType, icon)
        }
        else{
            console.log(`✅ Reusing icon for: ${iconType}`)
        }

        return this.icons.get(iconType)!
    }
}